{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "\n",
    "import plotly as plt\n",
    "\n",
    "pd.options.plotting.backend = \"plotly\"\n",
    "import plotly.io as pio\n",
    "pio.renderers.default = \"browser\"\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_429253/2437919683.py:7: DtypeWarning:\n",
      "\n",
      "Columns (1,2,5,6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "\n",
      "/tmp/ipykernel_429253/2437919683.py:15: DtypeWarning:\n",
      "\n",
      "Columns (1,2,5,6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "\n",
      "/tmp/ipykernel_429253/2437919683.py:23: DtypeWarning:\n",
      "\n",
      "Columns (1,2,5,6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>exon</th>\n",
       "      <th>cov_start</th>\n",
       "      <th>cov_end</th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100080</td>\n",
       "      <td>1453</td>\n",
       "      <td>1496</td>\n",
       "      <td>1373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100080</td>\n",
       "      <td>27100081</td>\n",
       "      <td>1454</td>\n",
       "      <td>1504</td>\n",
       "      <td>1376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100081</td>\n",
       "      <td>27100082</td>\n",
       "      <td>1463</td>\n",
       "      <td>1502</td>\n",
       "      <td>1384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100082</td>\n",
       "      <td>27100083</td>\n",
       "      <td>1467</td>\n",
       "      <td>1517</td>\n",
       "      <td>1390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100083</td>\n",
       "      <td>27100085</td>\n",
       "      <td>1472</td>\n",
       "      <td>1522</td>\n",
       "      <td>1404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100492</td>\n",
       "      <td>27100493</td>\n",
       "      <td>1117</td>\n",
       "      <td>1186</td>\n",
       "      <td>1088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100493</td>\n",
       "      <td>27100494</td>\n",
       "      <td>1116</td>\n",
       "      <td>1185</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100494</td>\n",
       "      <td>27100495</td>\n",
       "      <td>1111</td>\n",
       "      <td>1182</td>\n",
       "      <td>1076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100495</td>\n",
       "      <td>27100496</td>\n",
       "      <td>1107</td>\n",
       "      <td>1176</td>\n",
       "      <td>1075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100496</td>\n",
       "      <td>27100497</td>\n",
       "      <td>1109</td>\n",
       "      <td>1178</td>\n",
       "      <td>1056</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>402 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom exon_start  exon_end     gene   transcript exon cov_start  \\\n",
       "0    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100079   \n",
       "1    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100080   \n",
       "2    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100081   \n",
       "3    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100082   \n",
       "4    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100083   \n",
       "..     ...        ...       ...      ...          ...  ...       ...   \n",
       "397  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100492   \n",
       "398  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100493   \n",
       "399  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100494   \n",
       "400  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100495   \n",
       "401  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100496   \n",
       "\n",
       "      cov_end    S1    S2    S3  \n",
       "0    27100080  1453  1496  1373  \n",
       "1    27100081  1454  1504  1376  \n",
       "2    27100082  1463  1502  1384  \n",
       "3    27100083  1467  1517  1390  \n",
       "4    27100085  1472  1522  1404  \n",
       "..        ...   ...   ...   ...  \n",
       "397  27100493  1117  1186  1088  \n",
       "398  27100494  1116  1185  1084  \n",
       "399  27100495  1111  1182  1076  \n",
       "400  27100496  1107  1176  1075  \n",
       "401  27100497  1109  1178  1056  \n",
       "\n",
       "[402 rows x 11 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column = [\n",
    "            \"chrom\", \"exon_start\", \"exon_end\", \"gene\",\n",
    "            \"transcript\", \"exon\", \"cov_start\", \"cov_end\", \"cov\"\n",
    "        ]\n",
    "\n",
    "\n",
    "sample1 = pd.read_csv('/home/jethro/Projects/athena/test/annotated_beds/2302858-23074Z0010-1-BM-MPD-MYE-M-EGG2_S1_L001_markdup_annotated_bed.tsv.gz',names=column, sep='\\t')\n",
    "\n",
    "sample1 = sample1[sample1['gene']=='ANKRD26']\n",
    "sample1 = sample1[sample1['exon'] == 1]\n",
    "\n",
    "# sample1 = sample1.iloc[:10].reset_index(drop=True)\n",
    "sample1 = sample1.rename(columns={'cov': 'S1'})\n",
    "\n",
    "sample2 = pd.read_csv('/home/jethro/Projects/athena/test/annotated_beds/2302867-23075Z0089-1-PB-MPD-MYE-F-EGG2_S2_L001_markdup_annotated_bed.tsv.gz',names=column, sep='\\t')\n",
    "\n",
    "sample2 = sample2[sample2['gene']=='ANKRD26']\n",
    "sample2 = sample2[sample2['exon'] == 1]\n",
    "\n",
    "# sample2 = sample2.iloc[:10].reset_index(drop=True)\n",
    "sample2 = sample2.rename(columns={'cov': 'S2'})\n",
    "\n",
    "sample3 = pd.read_csv('/home/jethro/Projects/athena/test/annotated_beds/2302871-23075Z0092-1-BM-MPD-MYE-M-EGG2_S3_L001_markdup_annotated_bed.tsv.gz',names=column, sep='\\t')\n",
    "\n",
    "sample3 = sample3[sample3['gene']=='ANKRD26']\n",
    "sample3 = sample3[sample3['exon'] == 1]\n",
    "\n",
    "# sample3 = sample3.iloc[:10].reset_index(drop=True)\n",
    "sample3 = sample3.rename(columns={'cov': 'S3'})\n",
    "\n",
    "sample1.reset_index(drop=True, inplace=True)    \n",
    "sample2.reset_index(drop=True, inplace=True)\n",
    "sample3.reset_index(drop=True, inplace=True)\n",
    "\n",
    "\n",
    "samples = sample1.copy()\n",
    "samples['S2'] = sample2[['S2']]\n",
    "samples['S3'] = sample3['S3']\n",
    "\n",
    "samples.iloc[:402]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1913.398252184769"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[['S1', 'S2', 'S3']].stack().mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1520.7"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[['S1', 'S2', 'S3']].mean().mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>exon</th>\n",
       "      <th>cov_start</th>\n",
       "      <th>cov_end</th>\n",
       "      <th>S1_norm</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>mean-std</th>\n",
       "      <th>mean+std</th>\n",
       "      <th>S1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100080</td>\n",
       "      <td>1453</td>\n",
       "      <td>1496</td>\n",
       "      <td>1373</td>\n",
       "      <td>1441.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1379.0</td>\n",
       "      <td>1503.0</td>\n",
       "      <td>1888.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100080</td>\n",
       "      <td>27100081</td>\n",
       "      <td>1454</td>\n",
       "      <td>1504</td>\n",
       "      <td>1376</td>\n",
       "      <td>1445.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>1380.0</td>\n",
       "      <td>1510.0</td>\n",
       "      <td>1890.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100081</td>\n",
       "      <td>27100082</td>\n",
       "      <td>1463</td>\n",
       "      <td>1502</td>\n",
       "      <td>1384</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1390.0</td>\n",
       "      <td>1510.0</td>\n",
       "      <td>1901.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100082</td>\n",
       "      <td>27100083</td>\n",
       "      <td>1467</td>\n",
       "      <td>1517</td>\n",
       "      <td>1390</td>\n",
       "      <td>1458.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>1522.0</td>\n",
       "      <td>1907.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100083</td>\n",
       "      <td>27100085</td>\n",
       "      <td>1472</td>\n",
       "      <td>1522</td>\n",
       "      <td>1404</td>\n",
       "      <td>1466.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1407.0</td>\n",
       "      <td>1525.0</td>\n",
       "      <td>1913.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100494</td>\n",
       "      <td>27100495</td>\n",
       "      <td>1111</td>\n",
       "      <td>1182</td>\n",
       "      <td>1076</td>\n",
       "      <td>1123.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>1069.0</td>\n",
       "      <td>1177.0</td>\n",
       "      <td>1444.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100495</td>\n",
       "      <td>27100496</td>\n",
       "      <td>1107</td>\n",
       "      <td>1176</td>\n",
       "      <td>1075</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1067.0</td>\n",
       "      <td>1171.0</td>\n",
       "      <td>1439.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100496</td>\n",
       "      <td>27100497</td>\n",
       "      <td>1109</td>\n",
       "      <td>1178</td>\n",
       "      <td>1056</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>1441.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100497</td>\n",
       "      <td>27100498</td>\n",
       "      <td>1088</td>\n",
       "      <td>1155</td>\n",
       "      <td>1057</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1050.0</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>1414.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100498</td>\n",
       "      <td>27100499</td>\n",
       "      <td>1087</td>\n",
       "      <td>1150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1118.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>1163.0</td>\n",
       "      <td>1413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom exon_start  exon_end     gene   transcript exon cov_start  \\\n",
       "0    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100079   \n",
       "1    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100080   \n",
       "2    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100081   \n",
       "3    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100082   \n",
       "4    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100083   \n",
       "..     ...        ...       ...      ...          ...  ...       ...   \n",
       "399  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100494   \n",
       "400  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100495   \n",
       "401  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100496   \n",
       "402  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100497   \n",
       "403  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100498   \n",
       "\n",
       "      cov_end S1_norm    S2    S3    mean   std  mean-std  mean+std      S1  \n",
       "0    27100080    1453  1496  1373  1441.0  62.0    1379.0    1503.0  1888.9  \n",
       "1    27100081    1454  1504  1376  1445.0  65.0    1380.0    1510.0  1890.2  \n",
       "2    27100082    1463  1502  1384  1450.0  60.0    1390.0    1510.0  1901.9  \n",
       "3    27100083    1467  1517  1390  1458.0  64.0    1394.0    1522.0  1907.1  \n",
       "4    27100085    1472  1522  1404  1466.0  59.0    1407.0    1525.0  1913.6  \n",
       "..        ...     ...   ...   ...     ...   ...       ...       ...     ...  \n",
       "399  27100495    1111  1182  1076  1123.0  54.0    1069.0    1177.0  1444.3  \n",
       "400  27100496    1107  1176  1075  1119.0  52.0    1067.0    1171.0  1439.1  \n",
       "401  27100497    1109  1178  1056  1114.0  61.0    1053.0    1175.0  1441.7  \n",
       "402  27100498    1088  1155  1057  1100.0  50.0    1050.0    1150.0  1414.4  \n",
       "403  27100499    1087  1150   NaN  1118.0  45.0    1073.0    1163.0  1413.1  \n",
       "\n",
       "[404 rows x 16 columns]"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['mean'] = samples[['S1', 'S2', 'S3']].mean(axis=1).round()\n",
    "samples['std'] = samples[['S1', 'S2', 'S3']].std(axis=1).round()\n",
    "samples['mean-std'] = samples['mean'] - samples['std']\n",
    "samples['mean+std'] = samples['mean'] + samples['std']\n",
    "\n",
    "samples.rename(columns={'S1': 'S1_norm'},inplace=True)\n",
    "samples['S1'] = samples['S1_norm'] * 1.3\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom          object\n",
       "exon_start     object\n",
       "exon_end       object\n",
       "gene           object\n",
       "transcript     object\n",
       "exon           object\n",
       "cov_start      object\n",
       "cov_end        object\n",
       "S1_norm       float64\n",
       "S2             object\n",
       "S3             object\n",
       "mean          float64\n",
       "std           float64\n",
       "mean-std      float64\n",
       "mean+std      float64\n",
       "S1            float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = samples.astype({'S1': float, 'mean': float, 'S1_norm': float})\n",
    "samples.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples['assay'] = samples['mean'] * 1.05\n",
    "samples['assay-std'] = samples['mean-std'] * 1.05\n",
    "samples['assay+std'] = samples['mean+std'] * 1.05\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>exon</th>\n",
       "      <th>cov_start</th>\n",
       "      <th>cov_end</th>\n",
       "      <th>S1_norm</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>mean-std</th>\n",
       "      <th>mean+std</th>\n",
       "      <th>S1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100080</td>\n",
       "      <td>1453</td>\n",
       "      <td>1496</td>\n",
       "      <td>1373</td>\n",
       "      <td>1441.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1379.0</td>\n",
       "      <td>1503.0</td>\n",
       "      <td>1888.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100080</td>\n",
       "      <td>27100081</td>\n",
       "      <td>1454</td>\n",
       "      <td>1504</td>\n",
       "      <td>1376</td>\n",
       "      <td>1445.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>1380.0</td>\n",
       "      <td>1510.0</td>\n",
       "      <td>1890.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100081</td>\n",
       "      <td>27100082</td>\n",
       "      <td>1463</td>\n",
       "      <td>1502</td>\n",
       "      <td>1384</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1390.0</td>\n",
       "      <td>1510.0</td>\n",
       "      <td>1901.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100082</td>\n",
       "      <td>27100083</td>\n",
       "      <td>1467</td>\n",
       "      <td>1517</td>\n",
       "      <td>1390</td>\n",
       "      <td>1458.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>1522.0</td>\n",
       "      <td>1907.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100083</td>\n",
       "      <td>27100085</td>\n",
       "      <td>1472</td>\n",
       "      <td>1522</td>\n",
       "      <td>1404</td>\n",
       "      <td>1466.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1407.0</td>\n",
       "      <td>1525.0</td>\n",
       "      <td>1913.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100494</td>\n",
       "      <td>27100495</td>\n",
       "      <td>1111</td>\n",
       "      <td>1182</td>\n",
       "      <td>1076</td>\n",
       "      <td>1123.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>1069.0</td>\n",
       "      <td>1177.0</td>\n",
       "      <td>1444.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100495</td>\n",
       "      <td>27100496</td>\n",
       "      <td>1107</td>\n",
       "      <td>1176</td>\n",
       "      <td>1075</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1067.0</td>\n",
       "      <td>1171.0</td>\n",
       "      <td>1439.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100496</td>\n",
       "      <td>27100497</td>\n",
       "      <td>1109</td>\n",
       "      <td>1178</td>\n",
       "      <td>1056</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>1441.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100497</td>\n",
       "      <td>27100498</td>\n",
       "      <td>1088</td>\n",
       "      <td>1155</td>\n",
       "      <td>1057</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1050.0</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>1414.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>chr10</td>\n",
       "      <td>27100079</td>\n",
       "      <td>27100499</td>\n",
       "      <td>ANKRD26</td>\n",
       "      <td>NM_014915.3</td>\n",
       "      <td>1</td>\n",
       "      <td>27100498</td>\n",
       "      <td>27100499</td>\n",
       "      <td>1087</td>\n",
       "      <td>1150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1118.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>1163.0</td>\n",
       "      <td>1413.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom exon_start  exon_end     gene   transcript exon cov_start  \\\n",
       "0    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100079   \n",
       "1    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100080   \n",
       "2    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100081   \n",
       "3    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100082   \n",
       "4    chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100083   \n",
       "..     ...        ...       ...      ...          ...  ...       ...   \n",
       "399  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100494   \n",
       "400  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100495   \n",
       "401  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100496   \n",
       "402  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100497   \n",
       "403  chr10   27100079  27100499  ANKRD26  NM_014915.3    1  27100498   \n",
       "\n",
       "      cov_end S1_norm    S2    S3    mean   std  mean-std  mean+std      S1  \n",
       "0    27100080    1453  1496  1373  1441.0  62.0    1379.0    1503.0  1888.9  \n",
       "1    27100081    1454  1504  1376  1445.0  65.0    1380.0    1510.0  1890.2  \n",
       "2    27100082    1463  1502  1384  1450.0  60.0    1390.0    1510.0  1901.9  \n",
       "3    27100083    1467  1517  1390  1458.0  64.0    1394.0    1522.0  1907.1  \n",
       "4    27100085    1472  1522  1404  1466.0  59.0    1407.0    1525.0  1913.6  \n",
       "..        ...     ...   ...   ...     ...   ...       ...       ...     ...  \n",
       "399  27100495    1111  1182  1076  1123.0  54.0    1069.0    1177.0  1444.3  \n",
       "400  27100496    1107  1176  1075  1119.0  52.0    1067.0    1171.0  1439.1  \n",
       "401  27100497    1109  1178  1056  1114.0  61.0    1053.0    1175.0  1441.7  \n",
       "402  27100498    1088  1155  1057  1100.0  50.0    1050.0    1150.0  1414.4  \n",
       "403  27100499    1087  1150   NaN  1118.0  45.0    1073.0    1163.0  1413.1  \n",
       "\n",
       "[404 rows x 16 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['assay-std'], hoverinfo='skip',\n",
    "    mode='lines', name='assay -std dev',line=dict(color='rgba(197, 243, 243, 0.85)')))\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['assay+std'], fillcolor='rgba(197, 243, 243, 0.85)', hoverinfo='skip',\n",
    "    mode='lines', name='assay +std dev', fill='tonexty', line=dict(color='rgba(197, 243, 243, 1)')))\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['assay'], mode='lines', name='assay mean', line=dict(color='skyblue')))\n",
    "\n",
    "\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['mean-std'], hoverinfo='skip',\n",
    "    mode='lines', name='run -std dev',line=dict(color='rgba(83, 230, 123, 0.5)')))\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['mean+std'], fillcolor='rgba(83, 230, 123, 0.5)', hoverinfo='skip',\n",
    "    mode='lines', name='run +std dev', fill='tonexty', line=dict(color='rgba(83, 230, 123, 0)')))\n",
    "\n",
    "\n",
    "# fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['S1'], mode='lines', name='sample raw', line=dict(color='mediumturquoise')))\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['S1_norm'], mode='lines', name='sample', line=dict(color='red')))\n",
    "fig.add_trace(go.Scatter(x=samples['cov_start'], y=samples['mean'], mode='lines', name='run mean', line=dict(color='darkgreen')))\n",
    "fig.add_trace(go.Scatter(x=[samples['cov_start'].tolist()[0], samples['cov_start'].tolist()[-1]], y=[250]*len(samples['cov_start']), line=dict(color='red'), hoverinfo='skip', mode=\"lines\"))\n",
    "\n",
    "fig.update_layout(hovermode=\"x unified\")\n",
    "fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_429253/194208974.py:1: DtypeWarning:\n",
      "\n",
      "Columns (1,2,5,6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('/home/jethro/Projects/athena/test/annotated_beds/2302858-23074Z0010-1-BM-MPD-MYE-M-EGG2_S1_L001_markdup_annotated_bed.tsv.gz',names=column, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>exon</th>\n",
       "      <th>cov_start</th>\n",
       "      <th>cov_end</th>\n",
       "      <th>cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrom</td>\n",
       "      <td>exon_start</td>\n",
       "      <td>exon_end</td>\n",
       "      <td>gene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>cov_start</td>\n",
       "      <td>cov_end</td>\n",
       "      <td>cov</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1787325</td>\n",
       "      <td>1787442</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>NM_002074.5</td>\n",
       "      <td>11</td>\n",
       "      <td>1787325</td>\n",
       "      <td>1787326</td>\n",
       "      <td>1680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1787325</td>\n",
       "      <td>1787442</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>NM_002074.5</td>\n",
       "      <td>11</td>\n",
       "      <td>1787326</td>\n",
       "      <td>1787327</td>\n",
       "      <td>1712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1787325</td>\n",
       "      <td>1787442</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>NM_002074.5</td>\n",
       "      <td>11</td>\n",
       "      <td>1787327</td>\n",
       "      <td>1787328</td>\n",
       "      <td>1722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1787325</td>\n",
       "      <td>1787442</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>NM_002074.5</td>\n",
       "      <td>11</td>\n",
       "      <td>1787328</td>\n",
       "      <td>1787329</td>\n",
       "      <td>1727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224579</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134425195</td>\n",
       "      <td>134425335</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>NM_001015877.2</td>\n",
       "      <td>10</td>\n",
       "      <td>134425330</td>\n",
       "      <td>134425331</td>\n",
       "      <td>845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224580</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134425195</td>\n",
       "      <td>134425335</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>NM_001015877.2</td>\n",
       "      <td>10</td>\n",
       "      <td>134425331</td>\n",
       "      <td>134425332</td>\n",
       "      <td>834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224581</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134425195</td>\n",
       "      <td>134425335</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>NM_001015877.2</td>\n",
       "      <td>10</td>\n",
       "      <td>134425332</td>\n",
       "      <td>134425333</td>\n",
       "      <td>832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224582</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134425195</td>\n",
       "      <td>134425335</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>NM_001015877.2</td>\n",
       "      <td>10</td>\n",
       "      <td>134425333</td>\n",
       "      <td>134425334</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224583</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134425195</td>\n",
       "      <td>134425335</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>NM_001015877.2</td>\n",
       "      <td>10</td>\n",
       "      <td>134425334</td>\n",
       "      <td>134425335</td>\n",
       "      <td>828</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>224584 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  exon_start   exon_end  gene      transcript  exon  cov_start  \\\n",
       "0       chrom  exon_start   exon_end  gene      transcript  exon  cov_start   \n",
       "1        chr1     1787325    1787442  GNB1     NM_002074.5    11    1787325   \n",
       "2        chr1     1787325    1787442  GNB1     NM_002074.5    11    1787326   \n",
       "3        chr1     1787325    1787442  GNB1     NM_002074.5    11    1787327   \n",
       "4        chr1     1787325    1787442  GNB1     NM_002074.5    11    1787328   \n",
       "...       ...         ...        ...   ...             ...   ...        ...   \n",
       "224579   chrX   134425195  134425335  PHF6  NM_001015877.2    10  134425330   \n",
       "224580   chrX   134425195  134425335  PHF6  NM_001015877.2    10  134425331   \n",
       "224581   chrX   134425195  134425335  PHF6  NM_001015877.2    10  134425332   \n",
       "224582   chrX   134425195  134425335  PHF6  NM_001015877.2    10  134425333   \n",
       "224583   chrX   134425195  134425335  PHF6  NM_001015877.2    10  134425334   \n",
       "\n",
       "          cov_end   cov  \n",
       "0         cov_end   cov  \n",
       "1         1787326  1680  \n",
       "2         1787327  1712  \n",
       "3         1787328  1722  \n",
       "4         1787329  1727  \n",
       "...           ...   ...  \n",
       "224579  134425331   845  \n",
       "224580  134425332   834  \n",
       "224581  134425333   832  \n",
       "224582  134425334   830  \n",
       "224583  134425335   828  \n",
       "\n",
       "[224584 rows x 9 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2303914-23104Z0083-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303512-23093Z0108-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303500-23093Z0084-2-BMT-OTH-MYE-F-EGG2',\n",
       " '2303667-23095Z0107-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303062-23080Z0061-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303859-23104Z0023-2-CL-EQA-MYE-U-EGG2',\n",
       " '2303561-23093Z0115-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303860-23104Z0025-2-CL-EQA-MYE-U-EGG2',\n",
       " '2303204-23083Z0111-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303176-23082Z0039-2-BM-MPD-MYE-M-EGG2',\n",
       " '2301806-23026Z0058-2-CP-MPD-MYE-M-EGG2',\n",
       " '2303523-23093Z0121-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303614-23083Z0032-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303383-23089Z0052-2-UNK-LYM-MYE-M-EGG2',\n",
       " '2303203-23083Z0119-2-PB-CLL-MYE-M-EGG2',\n",
       " '2303183-23083Z0044-2-PB-MDS-MYE-F-EGG2',\n",
       " '2303233-23086Z0092-2-BM-MPD-MYE-M-EGG2',\n",
       " 'Oncospan-1778-1-AA1-BBB-MYE-U-EGG2',\n",
       " '2211998-22341Z0027-2-BM-ET-MYE-F-EGG2',\n",
       " '2303220-23074Z0084-2-DNAR-MPD-MYE-F-EGG2',\n",
       " '2303473-23090Z0076-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303209-23083Z0098-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303205-23083Z0110-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303730-23101Z0035-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303239-23086Z0080-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303521-23093Z0119-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303294-23087Z0106-2-PB-MPD-MYE-F-EGG2',\n",
       " '2303696-23096Z0113-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303689-23096Z0057-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303529-23090Z0070-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303177-23082Z0070-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303238-23086Z0077-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303308-23088Z0006-2-PB-MPD-MYE-M-EGG2',\n",
       " '2303975-23108Z0034-2-OTH-HNP-MYE-M-EGG2',\n",
       " '2303297-23087Z0092-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303242-23086Z0075-2-PB-MPD-MYE-F-EGG2',\n",
       " '2303225-23083Z0016-2-BM-MCL-MYE-F-EGG2',\n",
       " '2301759-23046Z0072-3-BM-MPD-MYE-F-EGG2',\n",
       " '2303206-23083Z0107-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303553-23094Z0036-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303237-23086Z0079-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303522-23093Z0110-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303207-23083Z0103-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303821-23103Z0032-2-BM-MPD-MYE-F-EGG2',\n",
       " '2303576-23094Z0103-2-BM-MPD-MYE-M-EGG2',\n",
       " '2303229-23083Z0012-2-BM-MPD-MYE-M-EGG2']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hsmetrics_paths = glob.glob('/home/jethro/Projects/athena/test/tri21/hs/*')\n",
    "hsmetrics = []\n",
    "\n",
    "for file in hsmetrics_paths:\n",
    "    hsmetrics.append(Path(file).stem.replace('_markdup.hsmetrics', ''))\n",
    "\n",
    "hsmetrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['2303207-23083Z0103-2-BM-MPD-MYE-M-EGG2', '2303553-23094Z0036-2-BM-MPD-MYE-M-EGG2', '2303205-23083Z0110-2-BM-MPD-MYE-F-EGG2', '2303308-23088Z0006-2-PB-MPD-MYE-M-EGG2', '2301759-23046Z0072-3-BM-MPD-MYE-F-EGG2', '2303203-23083Z0119-2-PB-CLL-MYE-M-EGG2', '2303473-23090Z0076-2-BM-MPD-MYE-M-EGG2', '2303209-23083Z0098-2-BM-MPD-MYE-F-EGG2', '2303614-23083Z0032-2-BM-MPD-MYE-F-EGG2', '2303294-23087Z0106-2-PB-MPD-MYE-F-EGG2', '2211998-22341Z0027-2-BM-ET-MYE-F-EGG2', '2303860-23104Z0025-2-CL-EQA-MYE-U-EGG2', '2303242-23086Z0075-2-PB-MPD-MYE-F-EGG2', '2303561-23093Z0115-2-BM-MPD-MYE-M-EGG2', '2303176-23082Z0039-2-BM-MPD-MYE-M-EGG2', '2303730-23101Z0035-2-BM-MPD-MYE-M-EGG2', '2303914-23104Z0083-2-BM-MPD-MYE-M-EGG2', '2303204-23083Z0111-2-BM-MPD-MYE-F-EGG2', '2303521-23093Z0119-2-BM-MPD-MYE-F-EGG2', '2303183-23083Z0044-2-PB-MDS-MYE-F-EGG2', '2303233-23086Z0092-2-BM-MPD-MYE-M-EGG2', '2303229-23083Z0012-2-BM-MPD-MYE-M-EGG2', '2303220-23074Z0084-2-DNAR-MPD-MYE-F-EGG2', '2303225-23083Z0016-2-BM-MCL-MYE-F-EGG2', '2303975-23108Z0034-2-OTH-HNP-MYE-M-EGG2', '2303206-23083Z0107-2-BM-MPD-MYE-F-EGG2', '2303237-23086Z0079-2-BM-MPD-MYE-F-EGG2', '2303859-23104Z0023-2-CL-EQA-MYE-U-EGG2', '2303529-23090Z0070-2-BM-MPD-MYE-F-EGG2', '2303576-23094Z0103-2-BM-MPD-MYE-M-EGG2', '2303062-23080Z0061-2-BM-MPD-MYE-M-EGG2', '2303821-23103Z0032-2-BM-MPD-MYE-F-EGG2', '2303512-23093Z0108-2-BM-MPD-MYE-M-EGG2', '2303667-23095Z0107-2-BM-MPD-MYE-M-EGG2', 'Oncospan-1778-1-AA1-BBB-MYE-U-EGG2', '2303297-23087Z0092-2-BM-MPD-MYE-M-EGG2', '2303239-23086Z0080-2-BM-MPD-MYE-M-EGG2', '2303522-23093Z0110-2-BM-MPD-MYE-F-EGG2', '2303238-23086Z0077-2-BM-MPD-MYE-M-EGG2', '2303523-23093Z0121-2-BM-MPD-MYE-F-EGG2', '2303696-23096Z0113-2-BM-MPD-MYE-F-EGG2', '2301806-23026Z0058-2-CP-MPD-MYE-M-EGG2', '2303177-23082Z0070-2-BM-MPD-MYE-M-EGG2', '2303689-23096Z0057-2-BM-MPD-MYE-F-EGG2', '2303383-23089Z0052-2-UNK-LYM-MYE-M-EGG2']\n"
     ]
    }
   ],
   "source": [
    "bed_paths = glob.glob('/home/jethro/Projects/athena/test/tri21/beds/*')\n",
    "beds = []\n",
    "\n",
    "for file in bed_paths:\n",
    "    beds.append(Path(file).stem.replace('_annotated.bed', ''))\n",
    "\n",
    "\n",
    "print(beds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils\n",
    "import load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {}\n",
    "\n",
    "for idx, bed in enumerate(bed_paths):\n",
    "    column = [\n",
    "            \"chrom\", \"exon_start\", \"exon_end\", \"gene\",\n",
    "            \"transcript\", \"exon\", \"cov_start\", \"cov_end\", \"cov\"\n",
    "        ]\n",
    "\n",
    "\n",
    "    sample = pd.read_csv(bed,names=column, sep='\\t', header=0, low_memory=False)\n",
    "\n",
    "    sample = sample[sample['gene']=='RUNX1']\n",
    "    sample = sample[sample['exon'] == 9]\n",
    "    sample = sample.rename(columns={'cov': f'S{idx+1  }'})\n",
    "\n",
    "    sample.reset_index(drop=True, inplace=True)   \n",
    "    # print(sample)\n",
    "\n",
    "    name = Path(bed).stem.replace('_annotated.bed', '')\n",
    "\n",
    "    sample = utils.unbin(sample)\n",
    "\n",
    "    dfs[name] = sample\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "hsmetrics_paths = glob.glob('/home/jethro/Projects/athena/test/tri21/hs/*')\n",
    "hsmetrics = {}\n",
    "\n",
    "for file in hsmetrics_paths:\n",
    "    name = Path(file).stem.replace('_markdup.hsmetrics', '')\n",
    "\n",
    "    df = load.LoadData().read_hsmetrics(file)\n",
    "    df.columns = df.iloc[0].tolist()\n",
    "    df.reset_index(inplace=True, drop=True)\n",
    "    df.drop(index=df.index[0], axis=0, inplace=True)\n",
    "    hsmetrics[name] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_hsmetrics = pd.concat(hsmetrics.values())\n",
    "all_hsmetrics\n",
    "all_hsmetrics = all_hsmetrics.astype({'ON_TARGET_BASES': int, 'PCT_USABLE_BASES_ON_TARGET': float, 'TOTAL_READS': int})\n",
    "# normalisation_value = (all_hsmetrics['ON_TARGET_BASES'] *\\\n",
    "#                  all_hsmetrics['PCT_USABLE_BASES_ON_TARGET']).sum()\n",
    "normalisation_value = all_hsmetrics['TOTAL_READS'].sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chrom  exon_start  exon_end   gene   transcript  exon  position    S1  \\\n",
      "0      21    34792129  34792615  RUNX1  NM_001754.5     9  34792129  1612   \n",
      "1      21    34792129  34792615  RUNX1  NM_001754.5     9  34792130  1625   \n",
      "2      21    34792129  34792615  RUNX1  NM_001754.5     9  34792131  1642   \n",
      "3      21    34792129  34792615  RUNX1  NM_001754.5     9  34792132  1659   \n",
      "4      21    34792129  34792615  RUNX1  NM_001754.5     9  34792133  1683   \n",
      "..    ...         ...       ...    ...          ...   ...       ...   ...   \n",
      "481    21    34792129  34792615  RUNX1  NM_001754.5     9  34792610  1812   \n",
      "482    21    34792129  34792615  RUNX1  NM_001754.5     9  34792611  1806   \n",
      "483    21    34792129  34792615  RUNX1  NM_001754.5     9  34792612  1802   \n",
      "484    21    34792129  34792615  RUNX1  NM_001754.5     9  34792613  1796   \n",
      "485    21    34792129  34792615  RUNX1  NM_001754.5     9  34792614  1788   \n",
      "\n",
      "       S1_norm    S2  ...   S41   S41_norm   S42   S42_norm   S43   S43_norm  \\\n",
      "0    31.233663  1417  ...  1668  43.542817  1706  33.793139  1650  35.857915   \n",
      "1    31.485547  1433  ...  1678  43.803865  1720  34.070457  1671  36.314289   \n",
      "2    31.814935  1454  ...  1700  44.378171  1739  34.446817  1691  36.748930   \n",
      "3    32.144322  1477  ...  1728  45.109106  1766  34.981643  1714  37.248768   \n",
      "4    32.609339  1494  ...  1755  45.813935  1780  35.258961  1734  37.683409   \n",
      "..         ...   ...  ...   ...        ...   ...        ...   ...        ...   \n",
      "481  35.108807  1508  ...  1946  50.799953  1989  39.398918  1934  42.029823   \n",
      "482  34.992553  1492  ...  1937  50.565010  1956  38.745240  1925  41.834234   \n",
      "483  34.915050  1489  ...  1934  50.486696  1954  38.705624  1921  41.747306   \n",
      "484  34.798796  1482  ...  1929  50.356172  1945  38.527348  1919  41.703842   \n",
      "485  34.643790  1464  ...  1905  49.729656  1906  37.754820  1895  41.182272   \n",
      "\n",
      "      S44   S44_norm  S45   S45_norm  \n",
      "0    1330  32.167859  589   9.411664  \n",
      "1    1339  32.385536  599   9.571454  \n",
      "2    1356  32.796704  611   9.763203  \n",
      "3    1381  33.401363  619   9.891035  \n",
      "4    1399  33.836717  628  10.034847  \n",
      "..    ...        ...  ...        ...  \n",
      "481  1447  34.997663  754  12.048208  \n",
      "482  1440  34.828358  746  11.920375  \n",
      "483  1436  34.731613  742  11.856459  \n",
      "484  1437  34.755799  737  11.776564  \n",
      "485  1420  34.344631  726  11.600794  \n",
      "\n",
      "[486 rows x 97 columns]\n"
     ]
    }
   ],
   "source": [
    "count = 1\n",
    "\n",
    "all_dfs = None\n",
    "all_dfs = list(dfs.values())[0].drop(['S1'], axis=1)\n",
    "\n",
    "\n",
    "for name, bed in dfs.items():\n",
    "    # print(df)\n",
    "    # bases = hsmetrics[name]['ON_TARGET_BASES']\n",
    "    # pct = hsmetrics[name]['PCT_USABLE_BASES_ON_TARGET']\n",
    "    \n",
    "    df = bed.copy()\n",
    "    hsmetrics[name] = hsmetrics[name].astype({\n",
    "            'ON_TARGET_BASES': int,\n",
    "            'PCT_USABLE_BASES_ON_TARGET': float,\n",
    "            'TOTAL_READS': int\n",
    "        })\n",
    "\n",
    "    # sample_bases = (\n",
    "    #         hsmetrics[name]['ON_TARGET_BASES'] * hsmetrics[name]['PCT_USABLE_BASES_ON_TARGET']\n",
    "    #     ).iloc[0] / normalisation_value\n",
    "\n",
    "    sample_bases = hsmetrics[name]['TOTAL_READS'].iloc[0] / normalisation_value\n",
    "\n",
    "    df[f'S{count}_norm'] = df[f\"S{count}\"] * sample_bases\n",
    "\n",
    "    all_dfs[f\"S{count}\"] = df[f\"S{count}\"]\n",
    "    all_dfs[f'S{count}_norm'] = df[f'S{count}_norm']\n",
    "\n",
    "    # print(name)\n",
    "    # print(df[f'S{count}'].mean())\n",
    "    # print(df[f'S{count}_norm'].mean())\n",
    "    # print(' ')\n",
    "    count += 1\n",
    "\n",
    "print(all_dfs)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [f\"S{x}\" for x in range(1,46)]\n",
    "all_dfs['mean'] = all_dfs[samples].mean(axis=1).round()\n",
    "all_dfs['std'] = all_dfs[samples].std(axis=1).round()\n",
    "all_dfs['3std'] = all_dfs['std'] * 3\n",
    "all_dfs['mean-3std'] = all_dfs['mean'] - all_dfs['3std']\n",
    "all_dfs['mean+3std'] = all_dfs['mean'] + all_dfs['3std']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [f\"S{x}_norm\" for x in range(1,46)]\n",
    "all_dfs['mean_norm'] = all_dfs[samples].mean(axis=1).round()\n",
    "all_dfs['std_norm'] = all_dfs[samples].std(axis=1).round()\n",
    "all_dfs['3std_norm'] = all_dfs['std_norm'] * 3\n",
    "all_dfs['mean-3std_norm'] = all_dfs['mean_norm'] - all_dfs['3std_norm']\n",
    "all_dfs['mean+3std_norm'] = all_dfs['mean_norm'] + all_dfs['3std_norm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# norm plot\n",
    "fig = go.Figure()\n",
    "\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean-3std_norm'], hoverinfo='skip',\n",
    "    mode='lines', name='-3 std dev',line=dict(color='rgba(197, 243, 243, 0.85)')))\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean+3std_norm'], fillcolor='rgba(197, 243, 243, 0.85)', hoverinfo='skip',\n",
    "    mode='lines', name='+3 std dev', fill='tonexty', line=dict(color='rgba(197, 243, 243, 1)')))\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean_norm'], mode='lines', name='mean', line=dict(color='skyblue')))\n",
    "\n",
    "\n",
    "\n",
    "for val in range(1, 46):\n",
    "\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=all_dfs['position'], y=all_dfs[f'S{val}_norm'],\n",
    "            mode='lines', name=f'S{val}'))\n",
    "\n",
    "\n",
    "\n",
    "fig.update_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = go.Figure()\n",
    "\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean-3std'], hoverinfo='skip',\n",
    "    mode='lines', name='-3 std dev',line=dict(color='rgba(197, 243, 243, 0.85)')))\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean+3std'], fillcolor='rgba(197, 243, 243, 0.85)', hoverinfo='skip',\n",
    "    mode='lines', name='+3 std dev', fill='tonexty', line=dict(color='rgba(197, 243, 243, 1)')))\n",
    "fig.add_trace(go.Scatter(x=all_dfs['position'], y=all_dfs['mean'], mode='lines', name='mean', line=dict(color='skyblue')))\n",
    "\n",
    "\n",
    "for val in range(1, 46):\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=all_dfs['position'], y=all_dfs[f'S{val}'],\n",
    "            mode='lines', name=f'S{val}'))\n",
    "\n",
    "\n",
    "fig.update_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 2303207-23083Z0103-2-BM-MPD-MYE-M-EGG2\n",
      "2 2303553-23094Z0036-2-BM-MPD-MYE-M-EGG2\n",
      "3 2303205-23083Z0110-2-BM-MPD-MYE-F-EGG2\n",
      "4 2303308-23088Z0006-2-PB-MPD-MYE-M-EGG2\n",
      "5 2301759-23046Z0072-3-BM-MPD-MYE-F-EGG2\n",
      "6 2303203-23083Z0119-2-PB-CLL-MYE-M-EGG2\n",
      "7 2303473-23090Z0076-2-BM-MPD-MYE-M-EGG2\n",
      "8 2303209-23083Z0098-2-BM-MPD-MYE-F-EGG2\n",
      "9 2303614-23083Z0032-2-BM-MPD-MYE-F-EGG2\n",
      "10 2303294-23087Z0106-2-PB-MPD-MYE-F-EGG2\n",
      "11 2211998-22341Z0027-2-BM-ET-MYE-F-EGG2\n",
      "12 2303860-23104Z0025-2-CL-EQA-MYE-U-EGG2\n",
      "13 2303242-23086Z0075-2-PB-MPD-MYE-F-EGG2\n",
      "14 2303561-23093Z0115-2-BM-MPD-MYE-M-EGG2\n",
      "15 2303176-23082Z0039-2-BM-MPD-MYE-M-EGG2\n",
      "16 2303730-23101Z0035-2-BM-MPD-MYE-M-EGG2\n",
      "17 2303914-23104Z0083-2-BM-MPD-MYE-M-EGG2\n",
      "18 2303204-23083Z0111-2-BM-MPD-MYE-F-EGG2\n",
      "19 2303521-23093Z0119-2-BM-MPD-MYE-F-EGG2\n",
      "20 2303183-23083Z0044-2-PB-MDS-MYE-F-EGG2\n",
      "21 2303233-23086Z0092-2-BM-MPD-MYE-M-EGG2\n",
      "22 2303229-23083Z0012-2-BM-MPD-MYE-M-EGG2\n",
      "23 2303220-23074Z0084-2-DNAR-MPD-MYE-F-EGG2\n",
      "24 2303225-23083Z0016-2-BM-MCL-MYE-F-EGG2\n",
      "25 2303975-23108Z0034-2-OTH-HNP-MYE-M-EGG2\n",
      "26 2303206-23083Z0107-2-BM-MPD-MYE-F-EGG2\n",
      "27 2303237-23086Z0079-2-BM-MPD-MYE-F-EGG2\n",
      "28 2303859-23104Z0023-2-CL-EQA-MYE-U-EGG2\n",
      "29 2303529-23090Z0070-2-BM-MPD-MYE-F-EGG2\n",
      "30 2303576-23094Z0103-2-BM-MPD-MYE-M-EGG2\n",
      "31 2303062-23080Z0061-2-BM-MPD-MYE-M-EGG2\n",
      "32 2303821-23103Z0032-2-BM-MPD-MYE-F-EGG2\n",
      "33 2303512-23093Z0108-2-BM-MPD-MYE-M-EGG2\n",
      "34 2303667-23095Z0107-2-BM-MPD-MYE-M-EGG2\n",
      "35 Oncospan-1778-1-AA1-BBB-MYE-U-EGG2\n",
      "36 2303297-23087Z0092-2-BM-MPD-MYE-M-EGG2\n",
      "37 2303239-23086Z0080-2-BM-MPD-MYE-M-EGG2\n",
      "38 2303522-23093Z0110-2-BM-MPD-MYE-F-EGG2\n",
      "39 2303238-23086Z0077-2-BM-MPD-MYE-M-EGG2\n",
      "40 2303523-23093Z0121-2-BM-MPD-MYE-F-EGG2\n",
      "41 2303696-23096Z0113-2-BM-MPD-MYE-F-EGG2\n",
      "42 2301806-23026Z0058-2-CP-MPD-MYE-M-EGG2\n",
      "43 2303177-23082Z0070-2-BM-MPD-MYE-M-EGG2\n",
      "44 2303689-23096Z0057-2-BM-MPD-MYE-F-EGG2\n",
      "45 2303383-23089Z0052-2-UNK-LYM-MYE-M-EGG2\n"
     ]
    }
   ],
   "source": [
    "for i, sample in enumerate(dfs.keys()):\n",
    "    print(i+1, sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_numeric(value:str) -> bool:\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    if not value:\n",
    "        return value\n",
    "    try:\n",
    "        if str(value)[0] == '0' and len(value) > 1 and not str(value)[1] == '.':\n",
    "            # catch things such as 01 which aren't real numbers\n",
    "            return value\n",
    "    except Exception:\n",
    "        # could raise lots of errors (i.e IndexErrors from slicing)\n",
    "        # lazily catch all and continue\n",
    "        return value\n",
    "\n",
    "    try:\n",
    "        float(value)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return value\n",
    "    \n",
    "    # it can float, check if it's an int or float\n",
    "    if '.' in str(value):\n",
    "        return float(value)\n",
    "    else:\n",
    "        return int(value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 <class 'str'>\n",
      "True <class 'bool'>\n",
      "01 <class 'str'>\n",
      "None <class 'NoneType'>\n",
      "test <class 'str'>\n",
      "0 <class 'int'>\n",
      "True <class 'bool'>\n"
     ]
    }
   ],
   "source": [
    "l = ['0', '0.1', '01', None, 'test', 0, 0.1]\n",
    "\n",
    "for val in l:\n",
    "    v = is_numeric(val)\n",
    "    print(v, type(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
