<!doctype html>
<html lang="en">
<head>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.10.24/css/jquery.dataTables.min.css">
	<link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/scroller/2.0.3/css/scroller.dataTables.min.css">
	<link rel="stylesheet" href="https://cdn.datatables.net/responsive/2.2.3/css/responsive.bootstrap4.min.css">

	<script type="text/javascript" language="javascript" src="https://code.jquery.com/jquery-3.5.1.js"></script>
	<script type="text/javascript" language="javascript" src="https://cdn.datatables.net/1.10.25/js/jquery.dataTables.min.js"></script>
	<script type="text/javascript" language="javascript" src="https://cdn.datatables.net/scroller/2.0.3/js/dataTables.scroller.min.js"></script>

  <style>
        /*  bootstrap passed as string from static/css*/ 
        $bootstrap

        body {
            padding-top: 100;
            padding-bottom: 150;
            width: 85%;
            margin-left: auto;
            margin-right: auto;
            font-size:18px;
            }

        .btn {
            margin-bottom: 12px;   
        }
        .summarybtn {
            transition-duration: 0.2s;
        }

        .summarybtn:hover {
            background-color: #07445c;
            color: white;
        }

        td {
            word-wrap: break-word;
            word-break:break-all
        }

        tr:hover {background-color:#ecebfc !important}

        /* styling for collapsible sections */
        .active, .collapsible:hover {
            background-color: rgb(255, 255, 255);
        }

        .collapsible:after {
            content: '+';
            color: rgb(255, 255, 255);
            font-weight: bold;
            float: right;
            margin-left: 5px;
        }

        .active:after {
            content: "-";
        }

        .content {
            padding: 0 18px;
            max-height: 0;
            overflow: hidden;
            transition: max-height 0.5s ease-out;
        }

        /* .fullPlots tbody {
            pointer-events: none;
        }             */

        .noHover{
            pointer-events: none;
        }

        div.dataTables_wrapper div.dataTables_scrollBody {
            /* without this it can collapse height on small rows 
            when searching if larger ones were in view */
            min-height: 600px;
            white-space: nowrap;
        }
    </style>
</head>

<body>
    <div class="card border-light mt-5 mb-5 p-3 bg-light table-responsive">
        <div class="card-header bg-info p-3 text-white">
            <h2>Coverage report for sample: $name </h2>
        </div><br>
        <div class="card-body bg-white">
            <h3>Report details</h3>
            <!--  Report details section -->
            <p>
            This report provides an assessment of coverage for $name.<br></br>
            It contains the following sections:
            <ul>
                <li> Summary including per gene coverage chart.</li>
                <li> Table of exons with sub-optimal coverage (<100% coverage at $threshold).</li>
                <li> Interactive plots of exons with sub-optimal coverage.</li>
                <li> A summary table of average coverage across all genes.</li>
                <li> Full gene coverage plots.</li>
                <li> Table of per exon coverage across all genes.</li>
                <li> Coverage of known variants (if specified).</li><br>
                n.b. all coverage plots are displayed relative to the <b>+ strand</b>.
            </ul>    
            </p>
        </div>

        <!-- Main body of report  -->
        <div class="card-body bg-white">
            <!--  Summary section -->
            <h2>Summary</h2> 
            <ul style="font-size:20px; line-height:2.0;">
                <!-- Include reference build if given -->
                $build
                $panel
                <li>Genes in the applied panel(s): <b>$total_genes</b></li>
                <li>Low coverage threshold: < <b>$threshold</b></li>
                <li>Panel coverage at $threshold: <b>$panel_pct_coverage %</b></li>
                <li>Genes with 100% coverage at $threshold: <b>$fully_covered_genes</b></li>
                <li>Genes <b>not</b> 100% covered at $threshold: <b>$gene_issues</b></li>
                $summary_text
            </ul>

            $summary_plot

            <br></br><br></br>
        
            <!-- Low covered exons table and plots -->
            <h2>Exons with &lt;100% coverage at $threshold</h2>
            <p style="font-size:18px">
                Of the <b>$total_genes</b> genes included in the panel, <b>$exon_issues</b> exons in 
                <b>$gene_issues</b> genes had sub optimal-coverage.
            </p>

            <div id="low_exons">
                <!-- Table for sub threshold exon coverage, headings passed from script, data in JS function -->
                <table id="low_exon_table" class="display nowrap" style="width:100%; text-align:center">

                </table>
            </div>
            <br></br>

            <br></br>

            To assess what portion of the exon(s) have sub-optimal coverage, the plots below display 
            an interactive window on hovering, showing the coverage at that current base.<br>
            The top right also contains a toolbar, with functions such as panning and zooming.

            $low_cov_plots
        
            <br></br><br></br>

            <div id="gene">
                <!-- Per gene table and plots -->
                <h2> Per gene coverage summary </h2>
                <br>
                The following section provides coverage metrics of each gene.<br>
                The table below contains metrics for each gene averaged across all of its exons. Plots of each gene are also included below.
                <br></br>

                <!-- Table for gene level coverage, headings passed from script, data in JS function -->
                <table id="gene_table" class="display nowrap" style="width:100%">
                    <thead>
                        <tr>
                            $gene_table_headings
                        </tr>
                    </thead>
                </table>

                <br>
                <!-- 'Table' for adding full gene plots to -->
                <table id="fullPlots" class='row-border nowrap noHover pageResize'>
                    <thead>
                        <tr>
                            <th></th>
                            <th></th>
                        </tr>
                    </thead>
                </table>

            </div>

            <br></br>

            <div id="exons">
                <!-- Table for gene level coverage, headings passed from script, data in JS function -->
                <h2> Coverage for all regions of all genes </h2>
                <br>
                The following section provides coverage metrics for each exon of each gene.
                <br></br>
                <table id="exon_table" class="display nowrap" style="width:100%">
                    <thead>
                        <tr>
                            $exon_table_headings
                        </tr>
                    </thead>
                </table>
            </div>
            <br></br>

            <!-- Optional section for displaying SNP information if generated -->
            <div id="snps">
                <h2> Coverage of Known Variants </h2>
                <br>
                Below are tables giving coverage of known variants. The low coverage table gives those not covered at $threshold, 
                and the high coverage table contains those covered above $threshold. <br></br>

                <li><b>$total_snps</b> variants were within the given panel regions and included for analysis</li>
                <li><b>$snps_covered</b> (<b>$snps_pct_covered %</b>) were covered at or above <b>$threshold</b></li>
                <li><b>$snps_not_covered</b> (<b>$snps_pct_not_covered %</b>) variants were <b>not</b> covered at <b>$threshold</b></li>
                <li><b>$snps_out_panel</b> (<b>$snps_pct_out_panel %</b>) variants spanned region boundaries </li>
                <br>
                $vcfs
                </br>
                <!-- snps_no_cov is div passed from script if variants spanning exon boundaries found -->
                $snps_no_cov
                Table of variants with low coverage (< $threshold) &nbsp
                <button class="btn btn-info collapsible btn-sm">Show / hide table</button>
                <div class="content">
                    <table class="snps" style="table-layout: fixed; width:100%">
                        $snps_low_cov
                    </table>
                </div>        
                <br>
                Table of variants with high coverage (>= $threshold) &nbsp
                <button class="btn btn-info collapsible btn-sm">Show / hide table</button>
                <div class="content">
                    <table class="snps" style="table-layout: fixed; width:100%">
                        $snps_high_cov
                    </table>
                </div>
                </br></br>
            </div>
        </div>
        <div class="card-footer bg-info p-1 text-white" style="display: flex; justify-content: space-between; align-items: baseline">
            <span>
                &nbsp&nbsp Report generated on $date
            </span>
            <span style="vertical-align:middle">
                Report created with Athena $version $logo
            </span>
        </div>    
    </body>
        
    <!-- DataTables function for full plots -->
    <script type="text/javascript" class="init">
        $(document).ready(function() {
            var data = [ $all_plots ];

            $('#fullPlots').DataTable( {
                data:           data,
                deferRender:    true,
                pageResize: true,
                info:       false, 
                scrollY:        "75vh",
                scrollCollapse: true,
                scroller:       true,
                scroller: {
                    loadingIndicator: true
                },
                'processing': true,
                'language': {
                    'loadingRecords': '&nbsp;',
                    'processing': 'Loading...'
                },
                "columnDefs": [
                    {
                        "targets": [ 0 ],
                        "visible": false,
                        "searchable": true
                    },
                    {
                        "targets": [ 1 ],
                        "visible": true,
                        "searchable": false,
                        "orderable": false
                    }
                ]
            });
        });
    </script>


    <!-- function to conditionally hide SNPs div if SNPs option not passed -->
    <script>
        // $total_snps is int passed from report script, set to JS snp_check var on writing to template to check against
        var snp_check = $total_snps
        if (snp_check == 0) { 
        document.getElementById('snps').style.display = "none";  
        }
    </script>

    <!-- function to make sections collapsible -->
    <script>
        var coll = document.getElementsByClassName("collapsible");
        var i;

        for (i = 0; i < coll.length; i++) {
        coll[i].addEventListener("click", function() {
            this.classList.toggle("active");
            var content = this.nextElementSibling;
            if (content.style.maxHeight){
            content.style.maxHeight = null;
            } else {
            content.style.maxHeight = content.scrollHeight * 2 + "px";
            } 
        });
        }
    </script>

    <!-- Function to copy text to clipboard, used for clinical summary text -->
    <script>
        function CopyToClipboard(id) {
            var r = document.createRange();
            r.selectNode(document.getElementById(id));
            window.getSelection().removeAllRanges();
            window.getSelection().addRange(r);
            document.execCommand('copy');
            window.getSelection().removeAllRanges();
        }
    </script>

    <!-- DataTables function for low exon coverage table -->
    <script type="text/javascript" class="init">
        function getColor(value){
            // generates value from 0-1 for colouring threshold column by value
            if (value >= 95) {
                var hue=((0 + (value/100))*80).toString(10);
            } else {
                var hue=((0 + (value/100))*67).toString(10);
            } 
            return ["hsl(",hue,",70%,50%)"].join("");
        }

        $(document).ready(function() {
            var data = $sub_threshold_stats ;
            var columnDefs = $low_exon_columns;
            var styleCol = columnDefs

            // get index of threshold column from list of columns
            // thresholdIdx = 0;
            var threshold = "$threshold";
            for(var k = 0; k < columnDefs.length; k++){
                if(columnDefs[k].title == threshold){
                    var thresholdIdx = k;
                }
            }

            $('#low_exon_table').DataTable( {
                data:           data,
                deferRender:    true,
                scrollY:        500,
                scrollCollapse: true,
                scroller:       true,
                lengthChange: false,
                columns: columnDefs,
                columnDefs: [{
                targets: thresholdIdx,
                render: function(data, type, row, meta) {
                    // colours threshold column by value
                    color = getColor(data)
                    barSize = 100 - data
                    return '<div style="background: linear-gradient(to right,' + color + ' ' + data +'%, #FFFFFF ' + barSize + '%);">' + data + '</div>';
                }
                }]
            });
        });
    </script>

    <!-- DataTables function for gene coverage table -->
    <script type="text/javascript" class="init">	
        $(document).ready(function() {
            var data = $gene_stats ;
            $('#gene_table').DataTable( {
                data:           data,
                deferRender:    true,
                scrollY:        500,
                scrollCollapse: true,
                scroller:       true,
                scroller: {
                    loadingIndicator: true
                },
                'processing': true,
                'language': {
                    'loadingRecords': '&nbsp;',
                    'processing': 'Loading...'
                }
            } );
        } );
    </script>

    <!-- DataTables function for exon coverage table -->
    <script type="text/javascript" class="init">	
        $(document).ready(function() {
            var data = $total_stats ;
            $('#exon_table').DataTable( {
                data:           data,
                deferRender:    true,
                scrollY:        500,
                scrollCollapse: true,
                scroller:       true,
                scroller: {
                    loadingIndicator: true
                },
                'processing': true,
                'language': {
                    'loadingRecords': '&nbsp;',
                    'processing': 'Loading...'
                }
            } );
        } );
    </script>
    </div>
</html>