<!doctype html>
<html lang="en">
    <head>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">

        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">


        <!-- local bootstrap -->
        <!-- <link href="../../resources/mdb/css/bootstrap.min.css" rel="stylesheet"> -->
        <!-- Material Design Bootstrap -->
        <!-- <link href="/resources/mdb/css/mdb.min.css" rel="stylesheet"> -->

        <style>
            body{
                padding-top: 100;
                padding-bottom: 150;
                width: 85%;
                margin-left: auto;
                margin-right: auto;
                font-size:18px
                }

            .cov-table {
                width: 100%;
                border: 10;
                table-layout:auto;
                overflow: auto;
                word-wrap:break-word;
                border-collapse:collapse;
                }   

            .snps {
                table-layout: fixed ;
                width: 100% ;
                }
            .snps th {
                width: 14%;
                text-align:left;
            }    
            .snps td {
                width: 14% ;
                text-align:left;
                }


            td {
                word-wrap: break-word;
                word-break:break-all
            }
            
            tr:hover {background-color:#ecebfc !important}

            /* styling for collapsible sections */

            .active, .collapsible:hover {
                background-color: rgb(255, 255, 255);
                }

            .collapsible:after {
                content: '+';
                color: rgb(255, 255, 255);
                font-weight: bold;
                float: right;
                margin-left: 5px;
                }

            .active:after {
                content: "-";
                }
            .content {
                padding: 0 18px;
                max-height: 0;
                overflow: hidden;
                transition: max-height 0.5s ease-out;
                /* background-color: #ffffff */
                } 
            .cov-table{
                color:black
            }
        </style>
    </head>

    <body>
        <div class="card border-light mt-5 mb-5 p-3 bg-light table-responsive">
        <div class="card-header bg-info p-3 text-white">
            <h2>Coverage report for sample: $name </h2>
        </div><br>
        <div class="card-body bg-white">
        
        <h2>Summary</h2>
        <br>
        
        <!--  Summary section -->
        <p style="font-size:18px">
        The following report provides an assessment of coverage for the given sample.<br></br>
        It contains the following sections: <br></br>
        <ul>
            <li> A table of exons with less than 100% coverage at $threshold .</li><br>
            <li> A series of interactive plots of exons with sub-optimal coverage.</li><br>
            <li> A summary table of coverage across all genes.</li><br>
            <li> A full table of per exon coverage across all genes.</li>
        </ul>    
        </p>
        <br></br>
        The plot below gives a summary of gene level coverage. Of the $total_genes genes in the panel, $gene_issues genes 
        had less than 100% coverage at the given threshold of $threshold . 

        $summary_plot
        <br></br>
        
        <!-- Low covered exons table and plots -->
        <h2>Exons with sub-optimal coverage</h2>
        <br>
        <p style="font-size:18px">
            Of the $total_genes genes included in the panel, $exon_issues exons in 
            $gene_issues genes had sub optimal coverage.<br>
            <br>
        </p>
        <table id=sub class="table table-striped cov-table sortable" width="100%">
            $sub_thrshld_stats
        </table>
        <br></br>
        To assess what portion of the exon(s) have sub-optimal coverage, the plots below display 
        an interactive window on hovering, showing the coverage at that current base. The top
        right also contains a toolbar, with functions such as panning and zooming.
        <br></br>
        $low_cov_plots
        <br>
        
        <div class="gene">
        <!-- Per gene table and plots -->
        <h2> Per gene coverage summary </h2>
        <br>
        The following section provides coverage metrics of each gene.<br>
        The table below contains metrics for each gene averaged across all of its exons. Plots of each gene are also included below.
        <br></br>

        <button type="button" class="btn btn-info btn-sm collapsible">Show / hide table</button>
        <div class="content table-responsive">
            <table class="table cov-table" width="100%">
                $gene_stats
            </table>
        </div>
        <br>
        <button class="btn btn-info btn-sm collapsible ">Show / hide plots</button>
        <div class="content" width="100%">
            $all_plots   
        </div>

    </div>
        <br></br>
    <div class="exons">
        <!-- Table of all exons in each gene -->
        <h2> Coverage for all regions of all genes </h2>
        <br>
        The following section provides coverage metrics for each exon of each gene.<br>
        This is a very large table with full comprehensive coverage metrics of all target regions.
        <br></br>
        <button class="btn btn-info btn-sm collapsible">Show / hide table</button>
        <div class="content">
        <br>
            <table class="cov-table" width="100%">
                $total_stats
            </table>
        </div>
    </div>
    <br></br>
    <!-- Optional section for displaying SNP information if generated -->
    <div class="snps">
        <h2> Coverage of SNPs </h2>
        <br>
        Below are two tables giving coverage of known SNPs. The first gives those not covered at the given threshold
        level, and the second contains those covered above the threshold. $total_snps SNPs were 
        within the given panel regions and included for analysis, of these $snps_not_covered SNPs were not covered at 
        $threshold, and $snps_covered were above $threshold .
        <br></br>
        Table of SNPs with low coverage &nbsp
        <button class="btn btn-info collapsible btn-sm">Show / hide table</button>
        <div class="content">
            <table class="snps" style="table-layout: fixed; width:100%">
                $snps_low_cov
            </table>
        </div>        
        <br>
        Table of SNPs with high coverage &nbsp
        <button class="btn btn-info collapsible btn-sm">Show / hide table</button>
        <div class="content">
            <table class="snps" style="table-layout: fixed; width:100%">
                $snps_high_cov
            </table>
        </div>
        </br></br>

        <div class="card-footer bg-info p-1 text-white">
            &nbsp&nbsp Report generated on $date
        </div>
    </div>
    
    <!-- function to conditionally display SNPs div if SNPs option has been passed -->
    <script>
        let divs = document.getElementsByClassName('snps');

        for (let x = 0; x < divs.length; x++) {
            let div = divs[x];
            let content = div.textContent.trim();
        
            if (content == '$snps_high_cov') {
            div.style.display = 'none';
        }
        }
    </script>


    <!-- function to make sections collapsible -->
    <script>
        var coll = document.getElementsByClassName("collapsible");
        var i;

        for (i = 0; i < coll.length; i++) {
        coll[i].addEventListener("click", function() {
            this.classList.toggle("active");
            var content = this.nextElementSibling;
            if (content.style.maxHeight){
            content.style.maxHeight = null;
            } else {
            content.style.maxHeight = content.scrollHeight + "px";
            } 
        });
        }
    </script>

    </div>  
    </div>

    </body>
</html>